Move tools dependencies into submodule, upgrade `golangci-lint`

guggero opened the pull request:
Replaces https://github.com/lightningnetwork/lnd/pull/6106.

It turns out extracting the tools into their own submodule has the advantage that we can upgrade `golangci-lint` without needing to upgrade some of our more critical dependencies (for example `google.golang.org/grpc`) in the main package as well.

NOTE: We disable most of the new linters that are now available in `golangci-lint` to keep the size of this PR reasonable. But the goal would be to enable some of them later on in individual PRs that also introduce the fixes for them.

naveensrinivasan commented:
> Replaces #6106.
> 
> It turns out extracting the tools into their own submodule has the advantage that we can upgrade `golangci-lint` without needing to upgrade some of our more critical dependencies (for example `google.golang.org/grpc`) in the main package as well.
> 
> NOTE: We disable most of the new linters that are now available in `golangci-lint` to keep the size of this PR reasonable. But the goal would be to enable some of them later on in individual PRs that also introduce the fixes for them.

Thanks for taking it over!

positiveblue reviewed (approved):
Nice! Thank you very much for taking the time to improve our suite for static analysis :tada:

It looks like there is a problem trying to compile one of the commits but everything else LGTM.

For the future, in case we need it, I wonder if there is an easy way to skip the checks on all the files with code generated automatically by tools like `grpc`, `mockgen`, `sqlc`, etc :thinking:

guggero commented:
Thanks for the review! I fixed the commit problem, some changes made it into the wrong commit.

I also fixed the Windows itest build by adding `build/dependencies.go`.

> For the future, in case we need it, I wonder if there is an easy way to skip the checks on all the files with code generated automatically by tools like grpc, mockgen, sqlc, etc.

That should be possible by adding a content based exclude rule!

yyforyongyu commented on .golangci.yml:
```diff
@@ -39,28 +55,73 @@ linters:
+  - varnamelen
+
+  # We want to allow TODOs.
+  - godox
```
Is this specific to VS code?

guggero replied:
Not sure what you mean. This is the description of the `godox` linter:
> Tool for detection of FIXME, TODO and other comment keywords

And since our code is full with TODOs, we don't want these warnings.

yyforyongyu commented on config.go:
```diff
@@ -1908,7 +1908,7 @@ func extractBitcoindRPCParams(networkName string,
 	}
 
-	chainDir := ""
+	var chainDir string
```
Ha interesting.

guggero replied:
Yeah, quite interesting. The linter now complains about "the initial value not being used".

yyforyongyu commented on contractcourt/nursery_store.go:
```diff
@@ -1058,52 +1058,52 @@ func (ns *NurseryStore) createHeightBucket(tx kvdb.RwTx,
 	byteOrder.PutUint32(heightBytes[:], height)
 
-	return chainBucket, hghtIndex, hghtIndex.NestedReadWriteBucket(
+	return hghtIndex, hghtIndex.NestedReadWriteBucket(
```
Didn't know we were not using it all the time😂

yyforyongyu commented on Makefile:
```diff
@@ -5,7 +5,7 @@ MOBILE_PKG := $(PKG)/mobile
 LINT_PKG := github.com/golangci/golangci-lint/cmd/golangci-lint
 GOACC_PKG := github.com/ory/go-acc
-GOIMPORTS_PKG := golang.org/x/tools/cmd/goimports
+GOIMPORTS_PKG := github.com/rinchsan/gosimports/cmd/gosimports
```
TIL. The only downside I see is that the project may go out of maintenance.

guggero replied:
True. But it seems the original one wasn't really that well maintained anyway.

yyforyongyu commented on docs/release-notes/release-notes-0.15.0.md:
```diff
@@ -98,6 +98,9 @@
   including new and deleted files.
 
+* [The `golangci-lint` package was updated and new linters were
+  enabled](https://github.com/lightningnetwork/lnd/pull/6244).
```
nit: we may want to mention for local dev we need to upgrade the `golangci-lint` or delete the old `golangci-lint`.

guggero replied:
Good point! Added that to the release notes.

yyforyongyu commented on routing/localchans/manager.go:
```diff
@@ -185,7 +185,7 @@ func (r *Manager) updateEdge(tx kvdb.RTx, chanPoint wire.OutPoint,
 	amtMin, amtMax, err := r.getHtlcAmtLimits(tx, chanPoint)
 	if err != nil {
-		return nil
+		return err
```
So this is wrong the whole time? We were not intentionally ignoring the error here right?

yyforyongyu reviewed (commented):
tACK. Thank you so much for fixing all these issues! Def have set an example for future contributors! No major blockings from me, just a few questions.

yyforyongyu reviewed (approved):
LGTM🥳
